interfaces/attrs.go

```diff
@@ -0,0 +1,197 @@
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package interfaces
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/snap"
+)
+
+type PlugData struct {
+	plug         *snap.PlugInfo
+	dynamicAttrs map[string]interface{}
+}
+
+type SlotData struct {
+	slot         *snap.SlotInfo
+	dynamicAttrs map[string]interface{}
+}
+
+type Attributes interface {
+	SetStaticAttr(key string, value interface{})
+	StaticAttr(key string) (interface{}, error)
+	StaticAttrs() map[string]interface{}
+	Attr(key string) (interface{}, error)
+	Attrs() (map[string]interface{}, error)
+	SetAttr(key string, value interface{}) error
+}
+
+func NewSlotData(slot *snap.SlotInfo, dynamicAttrs map[string]interface{}) *SlotData {
+	return &SlotData{
+		slot:         slot,
+		dynamicAttrs: dynamicAttrs,
+	}
+}
+
+func NewPlugData(plug *snap.PlugInfo, dynamicAttrs map[string]interface{}) *PlugData {
+	return &PlugData{
+		plug:         plug,
+		dynamicAttrs: dynamicAttrs,
+	}
+}
+
+func (attrs *PlugData) Interface() string {
+	return attrs.plug.Interface
+}
+
+func (attrs *PlugData) Name() string {
+	return attrs.plug.Name
+}
+
+func (attrs *PlugData) Ref() PlugRef {
+	return PlugRef{Snap: attrs.plug.Snap.Name(), Name: attrs.plug.Name}
+}
+
+func (attrs *PlugData) Snap() *snap.Info {
+	return attrs.plug.Snap
+}
+
+func (attrs *PlugData) Apps() map[string]*snap.AppInfo {
+	return attrs.plug.Apps
+}
+
+func (attrs *PlugData) SecurityTags() []string {
+	return attrs.plug.SecurityTags()
+}
+
+func (attrs *PlugData) StaticAttr(key string) (interface{}, error) {
+	if val, ok := attrs.plug.Attrs[key]; ok {
+		return val, nil
+	}
+	return nil, fmt.Errorf("attribute %q not found", key)
+}
+
+func (attrs *PlugData) SetStaticAttr(key string, value interface{}) {
+	if attrs.plug.Attrs == nil {
+		attrs.plug.Attrs = make(map[string]interface{})
+	}
+	attrs.plug.Attrs[key] = value
+}
+
+func (attrs *PlugData) StaticAttrs() map[string]interface{} {
+	return attrs.plug.Attrs
+}
+
+func (attrs *PlugData) Attr(key string) (interface{}, error) {
+	if attrs.dynamicAttrs != nil {
+		if val, ok := attrs.dynamicAttrs[key]; ok {
+			return val, nil
+		}
+	}
+	return attrs.StaticAttr(key)
+}
+
+func (attrs *PlugData) Attrs() (map[string]interface{}, error) {
+	if attrs.dynamicAttrs == nil {
+		return nil, fmt.Errorf("dynamic attributes not initialized")
+	}
+	return attrs.dynamicAttrs, nil
+}
+
+func (attrs *PlugData) SetAttr(key string, value interface{}) error {
+	if attrs.dynamicAttrs == nil {
+		return fmt.Errorf("dynamic attributes not initialized")
```

Comment:
Why not just initialize this?

Reply:
We want it to fail if you attempt to set dynamic attribute before hooks are even run. This will be initialized when after hooks are executed, so any AfterPreparePlug|Slot methods of interfaces will be able to call it without error.

Reply:
In that case it feels more like a panic

Reply:
my understanding is that this can be called also in the equivalent to Sanitize for setting defaults that is ok to override later

Reply:
to be clear I think to implement that 

                plugs:     make(map[string]map[string]*Plug),
                slots:     make(map[string]map[string]*Slot)

in repo needs to use *Data instead, then we would clone them before using them to make connections and then keep track of the final connection attributes with some object keeping those cloned/finalized *Data sides